### PR TITLE
[JENKINS-73835] Do not allow builds to be deleted while they are still running and ensure build discarders run after builds are fully complete

### PR DESCRIPTION
**Pull request by dwnusbaum**
See [JENKINS-73835](https://issues.jenkins.io/browse/JENKINS-73835) and https://github.com/jenkinsci/workflow-job-plugin/pull/470. I noticed this while investigating #9790, but it is a distinct issue.

This PR makes `Run.delete` throw an exception if it is called on a build which has not yet completed. It also adjusts some logic related to `LogRotator`, which is the main programmatic caller of `Run.delete` to avoid some race conditions and related issues. 

### Testing done

See new automated PRs.

### Proposed changelog entries

- Do not allow builds to be deleted while they are still building
- Ensure build discarders only process builds which have fully completed

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/java/hudson/model/Run.java
Comment (dwnusbaum) [inline core/src/main/java/hudson/model/Run.java:1891]: I moved this into `GlobalBuildDiscarderListener` so that `isLogUpdated() == true` when log rotation runs. It still runs synchronously, it just runs a little bit later now down in `onEndBuilding`. This allows us to check `isLogUpdated()` elsewhere to avoid race conditions with Pipeline builds.

File: core/src/main/java/hudson/tasks/LogRotator.java
Comment (dwnusbaum) [inline core/src/main/java/hudson/tasks/LogRotator.java:253]: This is to avoid race conditions involving Pipeline builds. Previously it was possible for log rotation to delete builds which had not yet fully completed, which meant that a `build.xml` file could be written back out into the build directory _after_ log rotation had deleted the build. I will try to demonstrate this downstream in `workflow-job` once I have an incremental build here.

File: core/src/main/java/jenkins/model/GlobalBuildDiscarderListener.java
Comment (dwnusbaum) [inline core/src/main/java/jenkins/model/GlobalBuildDiscarderListener.java:55]: This is a bit confusing, but essentially the old behavior was that `Run.run` called `Job.logRotate` while the build was in `POST_PRODUCTION` state, but then if the configuration of `GlobalBuildDiscarderConfiguration` included `JobGlobalBuildDiscarderStrategy` (default behavior), we also called `Job.logRotate` here while the build was in `COMPLETED` state. We only need it to be called once, and it's preferable to call it here so that we can check `Run.isLogUpdated` everywhere.

Also, this old double log rotation is why I am not concerned about removing [this call](https://github.com/jenkinsci/workflow-job-plugin/blob/d82e9f82c8115f04fa3f27fc38e18ca01c6410d2/src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java#L663-L669) in `workflow-job` and also why I am not worried about making the above call asynchronous. `BackgroundGlobalBuildDiscarder.processJob` has been calling `logRotate` synchronously for years now in default configurations, so I think that the asynchronous behavior in https://github.com/jenkinsci/workflow-job-plugin/commit/63fdbe8c2d0f74f97e8f3b92cc566e2fff52bdf3 is no longer critical.

File: test/src/test/java/hudson/model/RunTest.java
Comment (dwnusbaum) [inline test/src/test/java/hudson/model/RunTest.java:119]:
```suggestion
    @Issue("JENKINS-73835")
    @Test public void buildsMayNotBeDeletedWhileRunning() throws Exception {
```

File: test/src/test/java/hudson/tasks/LogRotatorTest.java
Comment (dwnusbaum) [inline test/src/test/java/hudson/tasks/LogRotatorTest.java:105]: Hmm, this was to check that the switch to `isLogUpdated` in `LogRotator.shouldKeepRun` did not break anything, since previously the call in `Run.run` would not have been able to delete the build with the switch. I guess though because of the redundant call via `GlobalBuildDiscarderListener` it would have worked anyway, there would just have been a log warning from the deleted code in `Run.run`. So maybe this test is no longer checking anything useful.

File: test/src/test/java/hudson/cli/DeleteBuildsCommandTest.java
Comment (dwnusbaum) [inline test/src/test/java/hudson/cli/DeleteBuildsCommandTest.java:142]: I did not notice this test before. It was added in https://github.com/jenkinsci/jenkins/pull/2310 along with all of the other tests for `delete-builds`.

I still think the previous behavior was undesirable - deleting builds that are still running and reachable in memory does not make sense, can lead to tons of errors in the Jenkins logs, and any builds deleted while running are very likely to come back after a Jenkins restart given that build completion will cause `build.xml` to be rewritten.

I can see the argument though that `Run.delete` should work like `Job.delete` and cancel the build if it is running, waiting up to 15 seconds and only throwing an exception if the build still hasn't completed at that point. @jglick suggested this approach as well, but I was hoping to avoid adding complexity here unless we really think that behavior is necessary. Any opinions on this?

File: core/src/main/java/hudson/tasks/LogRotator.java
Comment (jglick) [reply core/src/main/java/hudson/tasks/LogRotator.java:253]: Are you confident that this is safe for _freestyle_ builds? I.e., that `!logUpdated → !building` or conversely that `building → logUpdated`? The lifecycles for `AbstractBuild` and `WorkflowRun` are pretty different and the Javadoc has always been vague.

File: core/src/main/java/jenkins/model/GlobalBuildDiscarderListener.java
Comment (jglick) [reply core/src/main/java/jenkins/model/GlobalBuildDiscarderListener.java:55]: (https://github.com/jenkinsci/workflow-job-plugin/pull/70 for better linking)

File: test/src/test/java/hudson/cli/DeleteBuildsCommandTest.java
Comment (jglick) [reply test/src/test/java/hudson/cli/DeleteBuildsCommandTest.java:142]: No strong opinion. I think the proposed error behavior is intuitive enough. You can explicitly abort the build first if that is what you intended.

File: test/src/test/java/hudson/tasks/LogRotatorTest.java
Comment (jglick) [reply test/src/test/java/hudson/tasks/LogRotatorTest.java:105]: I guess, in principle, someone could want to keep zero builds if they were using publishers (GitHub Checks, Slack, etc.) to track failures?

**Review by jglick — APPROVED** [review]

File: core/src/main/java/hudson/tasks/LogRotator.java
Comment (dwnusbaum) [reply core/src/main/java/hudson/tasks/LogRotator.java:253]: I am pretty sure it is safe, but if you prefer we can switch to `isBuilding() || isLogUpdated()` like the check for the UI in Jelly (or maybe we should simplify that to only check `isLogUpdated` as well). See [here](https://github.com/jenkinsci/jenkins/blob/208238192f93e532284e4f8336e6a7849c50441b/core/src/main/java/hudson/model/Run.java#L518-L539) and the state enum [here](https://github.com/jenkinsci/jenkins/blob/208238192f93e532284e4f8336e6a7849c50441b/core/src/main/java/hudson/model/Run.java#L255-L278).

For `!logUpdated → !building`,`!isLogUpdated` is only true in `State.COMPLETED`, which ensures `!isBuilding` since `COMPLETED` is after `POST_PRODUCTION`.

The converse `building → logUpdated` should be ok as well, since if `isBuilding` then we must be in `State.CREATED` or `State.BUILIDING`, both of which are not `COMPLETED`, so `isLogUpdated` will be true.

For Pipeline things are simpler since `WorkflowRun.isLogUpdated` delegates to `WorkflowRun.isBuilding`.

Also just as a data point, if `isLogUpdated` could not be used on its own for non-pipelines, I think we would have seen it cause issues in tests due to the usage in [JenkinsRule.waitForCompletion](https://github.com/jenkinsci/jenkins-test-harness/blob/1542e2780eb84ab04b423a16b69da15ebea1e053/src/main/java/org/jvnet/hudson/test/JenkinsRule.java#L1633).

Comment (dwnusbaum) [reply core/src/main/java/hudson/tasks/LogRotator.java:253]: (And as far as I know the states for a `Run` can only move forward, although looking at the code I guess a subclass that calls some of the protected methods incorrectly could cause random state changes.)

Comment (jglick) [reply core/src/main/java/hudson/tasks/LogRotator.java:253]: > as far as I know the states for a `Run` can only move forward

That looks right.

> like the check for the UI in Jelly

Not sure what you mean. https://github.com/jenkinsci/jenkins/blob/208238192f93e532284e4f8336e6a7849c50441b/core/src/main/resources/hudson/model/Run/delete.jelly#L30 is only checking `building` and ought to be switch to only check `logUpdated`.

Comment (dwnusbaum) [reply core/src/main/java/hudson/tasks/LogRotator.java:253]: 🤦 I misread `keepLog` as `logUpdated`. I'll update that.

Comment (dwnusbaum) [reply core/src/main/java/hudson/tasks/LogRotator.java:253]: I fixed that in [f676be5](https://github.com/jenkinsci/jenkins/pull/9810/commits/f676be52239c943c2b43e9869ca923e63658d9b0).

**Review by jglick — APPROVED** [review]

**Review by scherler — APPROVED** [review]

**Comment by dwnusbaum** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by basil** [comment]
```
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 7.344 s <<< FAILURE! -- in org.jenkins.plugins.lockableresources.LockStepTest
[ERROR] org.jenkins.plugins.lockableresources.LockStepTest.deleteRunningBuildNewBuildClearsLock -- Time elapsed: 7.330 s <<< ERROR!
java.io.IOException: Unable to delete p #1 because it is still running
	at hudson.model.Run.delete(Run.java:1555)
	at org.jenkins.plugins.lockableresources.LockStepTest.deleteRunningBuildNewBuildClearsLock(LockStepTest.java:725)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:658)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:840)
```

**Comment by dwnusbaum** [comment]
I'll take a look at `LockStepTest.deleteRunningBuildNewBuildClearsLock`. Perhaps the test can just be deleted now.

**Comment by dwnusbaum** [comment]
https://github.com/jenkinsci/lockable-resources-plugin/pull/716 deletes the test.

**Comment by basil** [comment]
Ever since this PR was merged, `hudson.cli.DeleteBuildsCommandTest#deleteBuildsShouldFailIfTheBuildIsRunning` has been consistently failing on Windows CI builds with:

```
java.nio.file.DirectoryNotEmptyException: C:\Jenkins\agent\workspace\Core_jenkins_master\test\target\j h14545459714006339336\jobs\aProject\builds\1
	at java.base/sun.nio.fs.WindowsFileSystemProvider.implDelete(WindowsFileSystemProvider.java:272)
	at java.base/sun.nio.fs.AbstractFileSystemProvider.deleteIfExists(AbstractFileSystemProvider.java:110)
	at java.base/java.nio.file.Files.deleteIfExists(Files.java:1191)
	at org.jvnet.hudson.test.TemporaryDirectoryAllocator.delete(TemporaryDirectoryAllocator.java:146)
	at org.jvnet.hudson.test.TemporaryDirectoryAllocator.delete(TemporaryDirectoryAllocator.java:136)
	at org.jvnet.hudson.test.TemporaryDirectoryAllocator.delete(TemporaryDirectoryAllocator.java:136)
	at org.jvnet.hudson.test.TemporaryDirectoryAllocator.delete(TemporaryDirectoryAllocator.java:136)
	at org.jvnet.hudson.test.TemporaryDirectoryAllocator.delete(TemporaryDirectoryAllocator.java:136)
	at org.jvnet.hudson.test.TemporaryDirectoryAllocator.dispose(TemporaryDirectoryAllocator.java:104)
	at org.jvnet.hudson.test.TestEnvironment.dispose(TestEnvironment.java:84)
	at org.jvnet.hudson.test.JenkinsRule.after(JenkinsRule.java:538)
	at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:676)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:840)
	Suppressed: java.io.IOException: These files still exist : log
		at org.jvnet.hudson.test.TemporaryDirectoryAllocator.delete(TemporaryDirectoryAllocator.java:150)
		... 10 more
```

First failure was https://ci.jenkins.io/job/Core/job/jenkins/job/master/6632/testReport/junit/hudson.cli/DeleteBuildsCommandTest/windows_jdk17___Windows___JDK_17___Build___Test___deleteBuildsShouldFailIfTheBuildIsRunning/

**Comment by dwnusbaum** [comment]
Yes, looks like the test needs to forcibly stop the build. I will file a PR to fix it in a bit.

**Comment by dwnusbaum** [comment]
I filed https://github.com/jenkinsci/jenkins/pull/9876 to stabilize that test.

File: test/src/test/java/hudson/tasks/LogRotatorTest.java
Comment (jglick) [inline test/src/test/java/hudson/tasks/LogRotatorTest.java:112]: Failure in https://github.com/jenkinsci/jenkins/pull/9921/checks?check_run_id=32166988156 (#9921) looks like a flake? Missing `await`?

Comment (dwnusbaum) [reply test/src/test/java/hudson/tasks/LogRotatorTest.java:112]: I don't know what we would `await` for, because log rotation runs synchronously during build completion. Unless this is a test-specific timing issue due to `JenkinsRule.buildAndAssertSuccess` just waiting for `QueueTaskFuture.get` instead of `!Run.isLogUpdated`, then this test probably just needs to be skipped on Windows.

Comment (dwnusbaum) [reply test/src/test/java/hudson/tasks/LogRotatorTest.java:112]: From some testing, `j.buildAndAssertStatus(Result.SUCCESS, p)` is guaranteed to block until `GlobalBuildDiscarderListener.onFinalized` has completed. Perhaps the `BuildWatcher` background thread tried to copy the logs at an inopportune time and caused log rotation to fail. Seems best to delete the `BuildWatcher` here in case it made any of the other tests flaky on Windows and then probably skip this test explicitly on Windows as well.

Comment (dwnusbaum) [reply test/src/test/java/hudson/tasks/LogRotatorTest.java:112]: Also the `AccessDeniedException` seems strange. Either way, I filed https://github.com/jenkinsci/jenkins/pull/9923.

File: core/src/main/java/jenkins/model/GlobalBuildDiscarderListener.java
Comment (IppX) [reply core/src/main/java/jenkins/model/GlobalBuildDiscarderListener.java:55]: Hi, after some digging into the behavior of globalBuildDiscarder I ended here.

As a Jenkins admin, I expected the specific (or simple) global discarder to take precedence over any project build discarder, especially if I remove the global project build discarder. This is also what I implied from reading the descriptions [here](https://github.com/jenkinsci/jenkins/blob/365d03ec1d97d2d962baf9a7e4dedf1697535ebe/core/src/main/resources/jenkins/model/JobGlobalBuildDiscarderStrategy/config.properties#L1) and [here](https://github.com/jenkinsci/jenkins/blob/365d03ec1d97d2d962baf9a7e4dedf1697535ebe/core/src/main/resources/jenkins/model/SimpleGlobalBuildDiscarderStrategy/config.properties#L1) (even though it's not explicitly mentioned there and does not contradict the actual behavior). I also found multiple occurrences on SO and forums mentioning that global specific job discarder will take precedence, which is not the case. This is now obvious with the change line 51 here but was already the behavior before as I tested with 2.462.2.

What makes it even more confusing, is the way the discarders are merged and applied. The most aggressive policy takes over. This is confirmed to be an expected behavior from the [tests](https://github.com/jenkinsci/jenkins/blob/365d03ec1d97d2d962baf9a7e4dedf1697535ebe/test/src/test/java/jenkins/model/GlobalBuildDiscarderTest.java#L115-L126). However, from an admin perspective I'd see the least aggressive discarder to be a safer approach (e.g. for compliance reasons).

I'm curious however, to get your input on the topic. Also let me know if I should raise this somewhere else.

For context, this investigation started as a result of the security eng. telling me that he can erase his traces by overriding build discarder in a job. Example: normal run -> evil run (set buildDiscarder to 1) -> normal run. Then we can't see what happened in the evil run anymore. We also have some compliance policy that require us to keep production pipelines history up to a year which we would like to enforce.

Comment (jglick) [reply core/src/main/java/jenkins/model/GlobalBuildDiscarderListener.java:55]: I think the current behavior is as designed; build discarders are mainly meant to trim disk space, so it would be normal to have a blanket policy that builds over a month old are not kept, while a particular job that runs every five minutes for some automation is configured to discard all but the last build since history is trash. If your interest is in auditing, you would better use one of the many plugins which stream events or even whole build logs from Jenkins to external systems, which could be configured to only accept “create/append”-type operations and reject attempts to delete anything even if the Jenkins controller were to be compromised somehow. After all, even without any discarders, your white-hat engineer could simply delete the job after one build, along with all of its builds. You could also make the entire controller view-only to regular developers, using various “as code” systems.

Comment (daniel-beck) [reply core/src/main/java/jenkins/model/GlobalBuildDiscarderListener.java:55]: @IppX 

> This is also what I implied from reading the descriptions [here](https://github.com/jenkinsci/jenkins/blob/365d03ec1d97d2d962baf9a7e4dedf1697535ebe/core/src/main/resources/jenkins/model/JobGlobalBuildDiscarderStrategy/config.properties#L1) and [here](https://github.com/jenkinsci/jenkins/blob/365d03ec1d97d2d962baf9a7e4dedf1697535ebe/core/src/main/resources/jenkins/model/SimpleGlobalBuildDiscarderStrategy/config.properties#L1) (even though it's not explicitly mentioned there and does not contradict the actual behavior). I also found multiple occurrences on SO and forums mentioning that global specific job discarder will take precedence, which is not the case. … The most aggressive policy takes over

In addition to what Jesse wrote, for global build discarders, this seems to be reflected in the difference between "Default Build Discarder" and "Specific Build Discarder" for Global Build Discarders. Compare

> Build discarder will be applied for jobs without a specific build discarder, after any build finishes. This makes it possible to set a sensible default and override the discarder for specific jobs, that needs a longer retention policy.

and

> The selected build discarder will be applied after any build finishes, as well as periodically.

"Precedence" perhaps in the sense that, if it's more restrictive (less retention time/fewer retained builds) you end up keeping fewer builds.

**Pull request merged**
